### ISSUE TITLE
Why are subclasses of the user-selected checker classes also run?

### ISSUE THREAD

**Issue by mhidas** [ioos/compliance-checker#295]
The `_get_valid_checkers` method in ([suite.py](https://github.com/ioos/compliance-checker/blob/master/compliance_checker/suite.py#L88)) is supposed to filter out, from a list of checkers selected by the user, those checker classes that are appropriate for the dataset being checked. However, it also sneakily adds any subclasses of those checkers, leading to unexpected results.

On [line 109](https://github.com/ioos/compliance-checker/blob/master/compliance_checker/suite.py#L109-L113):

``` python
            # add any subclasses of the checker class
            for subc in a.__subclasses__():
                if subc not in all_checked:
                    all_checked.add(subc)
                    checker_queue.add((name, subc))
```

I am in the process of creating version 1.4 of the IMOS checker, which I was hoping to be simply a subclass of the v1.3 checker class. However, with the code above, if I do this I can no longer run just the v1.3 checks, because v1.4 is then automatically selected.

I could avoid this by following the example in acdd.py (with the different versions being subclasses of a common base class). I just wonder whether there's a use case that makes the above behaviour desirable?

**Comment by kwilcox** [comment]
Valid point @mhidas. I have actually consider this a "feature" since we can subclass something like "CF1.6" and implement very specific checks in addition to those in "CF1.6".  Maybe it's time to consider the ability to run more than one checker at once via all interfaces?

**Comment by lukecampbell** [comment]
@daf do you recall why we took this approach? It's been a while since I dug into the checker selection mechanics.

**Comment by daf** [comment]
It's rooted in being able to support multiple types of incoming data sources and apply the same checks to both: netcdf files/opendap urls AND sos gc/ds xml files/urls.  The idea was to define your checks in a Base class, and have a stub class inheriting from both your specific checker's Base and the BaseNCCheck/BaseSOSGCCheck/BaseSOSDSCheck classes to pair them up with sources.  The IOOS metadata checker is the only one still operating in that fashion, CF and ACDD appear to have moved on (CF wouldn't really make sense there anyway but ACDD has a case, it used to work).

There was also the `register_checker` concept that got added during plugin work by @fxmzb123 but now I don't see that being used in the discovery logic.  I don't remember what happened there.

@mhidas, I'm not sure I understand the issue - how are you selecting the checker to run on the command line?  The versioning capability (`--check your_checker:1.3`) should be handling this, or I'm not understanding what's going on..

**Comment by mhidas** [comment]
@daf I think won't have a problem once I rewrite my checkers the way it's done in ACDD (as you described above). I was just not aware of the intention to use this mechanism to support multiple data formats per checker. It's not really relevant for us, because our convention is specific to NetCDF files.

What I was trying to, thinking it's a bit simpler, was:
```python
class IMOSCheck(BaseNCCheck):
    """Compliance-checker check suite for the IMOS netcdf conventions
    """
    register_checker = True
    _cc_spec = 'imos'
    _cc_spec_version = '1.3'
    ...

class IMOS1_4Check(IMOSCheck):
    """Compliance-checker check suite for the IMOS netcdf conventions v1.4
    """
    _cc_spec_version = '1.4'
```
and in setup.py:
```python
    entry_points         = {
        'compliance_checker.suites': [
            'imos13 = cc_plugin_imos.imos:IMOSCheck'
            'imos14 = cc_plugin_imos.imos:IMOS1_4Check'
        ]
    }
```
With this setup, even if I specify `--test imos:1.3` on the command line, the code I referred to in the initial post makes sure that the `IMOS1_4Check` is also added to the list of checkers applied.

**Comment by daf** [comment]
Ok, I think I understand.  I think your setup should work and I'm in favor of the way you've laid your code out - it shouldn't force you to restructure so that your IMOS checks are in a shared base class, you should be able to inherit as you've laid out and it'll "just work".  Will need to think about a solution - maybe go down the subclasses until certain `_cc_*` metadata attributes are discovered, indicating a "top level" checker?

**Comment by mhidas** [comment]
Yep, that could work... or until the attribute `register_checker = True`  is found?

**Comment by benjwadams** [comment]
Closing due to changes implemented in #667.  SOS checker is implemented as a special case where subclasses are run, but by default, checkers do not use subclass behavior.
